Launch http bridge and wallet together

Anviking opened the pull request:
# Issue Number

#8 


# Overview

- [x] I have created a new `CLI` module for logic to be shared between multiple executables
- [x] I have added a new `cardano-wallet-launcher` executable
- [x] **WIP** Launcher tests (I'm seeing some strange things)

# Comments
- I have now rebased into 3 commits. **View by commit might be useful**

- The wallet crashes when it's trying to connect to the node that is still launching 😞 . If we only need 5 lines of retry mechanism it should be simple though.

- While I see strange things in unit tests / ghci, I also believe this is working. (from testing with `stack exec`)

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Refactor into new CLI module

To be shared between all executables (cardano-wallet-server and
cardano-wallet-launcher), and by providing an @Encodable@ class, should
make it easy for the launcher to forward arguments to the server.

(launcher is introduced in a later commit)

Anviking commented on app/server/Launcher.hs:
```diff
@@ -0,0 +1,45 @@
+bracketProcess :: CreateProcess -> IO () -> IO ()
+bracketProcess process action = do
+    _ <- withCreateProcess
+        process {delegate_ctlc = True, std_out = CreatePipe }
```
With this `delegate_ctlc`, `ctrl-d` works. `ctrl-c` still doesn't work…

I'd also like to redirect std_out, but don't know how yet.

Anviking commented on app/server/Launcher.hs:
```diff
@@ -0,0 +1,45 @@
+        -- if this IO action throws, @withCreateProcess@ will kill the node and
+        -- rethrow.
+        concurrently action (node ph)
+    return ()
```
This all is fairly intricate, and I wonder if one should have another approach. `withCreateProcess ` doesn't do everything we need. 

Also wonder if one can have some state-machine tests for this easily.

Should be able to do some unit tests at least.

Anviking replied:
Wait this _is_ wrong. Let's just not use `withCreateProcess`.

KtorZ commented on app/server/Launcher.hs:
```diff
@@ -0,0 +1,45 @@
+-- exception is raised during the computation. If an exception is raised, then
+-- bracket will re-raise the exception (after performing the release).
+--
+-- http://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception-Base.html#v:bracket
```
The `bracket` pattern is very common pattern in Haskell. I wouldn't bother putting that as documentation in the module. Or maybe, just a link to an article but not an entire paragraph for sure :+1:

KtorZ commented on app/server/Launcher.hs:
```diff
@@ -0,0 +1,45 @@
+
+import Control.Concurrent.Async
+import Prelude
+import System.Process
```
We'll favor explicit imports here :+1:

KtorZ commented on app/server/Launcher.hs:
```diff
@@ -0,0 +1,45 @@
+-- Both are expected to run forever.
+--
+-- Sadly you cannot quit this in ghci
+bracketProcess :: CreateProcess -> IO () -> IO ()
```
It's gonna be a bit "impractical" for users of that function to start a process because they'll have to deal with the internals of `CreateProcess` as well. Plus, this starts a process and, runs an IO action in the current process.  In practice, we want the launcher to start both the cardano-http-bridge and the wallet-server. This seemingly means that we need:

- An executable / CLI for the bridge
- An executable / CLI for the wallet

And, run both of them in a separate process, keeping the launcher thread as a "supervisor". So, the API here could be much more like:

```hs
data Command = Command
    { cmdName :: String
    , cmdArgs :: [String]
    }

launch :: [Command] -> IO [ProcessHandle]
```

What we do with the handles and what are the action isn't relevant for now. There's no need to "leak" logic about the node or the bridge within this function. For stdout, there a choice to make. We can either only bind the output of the first command, a bit arbitrarily (and stating it in the function's doc), or bind the output of all the commands, or, leave the decision to the caller and have a flag in the `Command` representation.

Anviking commented on app/launcher/Main.hs:
```diff
@@ -0,0 +1,102 @@
+|]
+
+
+type Process = (Maybe Handle, Maybe Handle, Maybe Handle, ProcessHandle)
```
I did this for [`cleanupProcess`](http://hackage.haskell.org/package/process-1.6.5.0/docs/System-Process.html#v:cleanupProcess). Not sure if it is needed.

KtorZ replied:
What you probably _really want_ here is a record :)

Anviking commented on app/launcher/Main.hs:
```diff
@@ -0,0 +1,102 @@
+
+    forM_ processes (forkIO . monitorAndKill processes)
+
+    monitorAndKill processes (head processes)
```
I did this to get the main thread to wait, which is, I'll admit a bit sketchy.

Anviking replied:
@KtorZ Could you have a look at the sanity of this `monitorAndKill` approach in particular? 🙏 

I fork threads to run `waitForProcess` on each process, and kill everything if one exits.

KtorZ replied:
Looking at the PR now :eyes:

KtorZ replied:
This doesn't feel right. `monitorAndKill` waits for the first process to die and kill the others if something happens :thinking: ... What if the first process dies? The others will keep on running.
I'd advise to have a look at `Concurrently` and `mapConcurrently` from the `async` package; those are useful utils to do easy "supervision" of process or, at the very least, ensure that they are either all up, or all dead.

Anviking replied:
A problem with `mapConcurrently` is that a process could exit successfully, without killing the others.

_ah, right, that is easy to fix_ 🙃

`monitorAndKill` is agnostic to this.

> What if the first process dies? The others will keep on running.

I'm actually not quite sure what you mean here. It contradicts 

> waits for the first process to die and kill the others if something happens

Previously, I ran `monitorAndKill` once more from the main thread waiting for `head processes` instead of just `forever $ threadDelay maxBound`. Might have been very confusing — sorry.

_will look into testing this regardless_

Anviking commented on app/launcher/wallet.sh:
```diff
@@ -0,0 +1,6 @@
+
+#while true; do echo "I'm a wallet"; sleep 1; done
+echo "I'm a wallet"
+sleep 20
```
Idea was to (at least) have temporary manual tests using files like this 🤔

KtorZ replied:
Oui :+1:

KtorZ commented on app/cli/CLI.hs:
```diff
@@ -0,0 +1,75 @@
+    -> Docopt
+    -> Option
+    -> (String -> Either String a)
+    -> IO a
```
What about `Decodable a => Arguments -> Docopt -> Option -> IO a` ?

Anviking replied:
I'm slightly sceptical towards the inflexibility of type-classes. Here there is no need.

It also cause type-inference problems with 
```haskell
    let getArg' = getArg args cli
```

**But**, if we were to make `getArg` more magical anyway,
```haskell
getArg :: Decodable a => Option -> ReaderT (Arguments, Docopt) IO a
```
, and hide 

```haskell
    args <- parseArgsOrExit cli =<< getArgs
    when (args `isPresent` (longOption "help")) $ exitWithUsage cli
```

in an extended `runReaderT`,

I'd do it. Should we?

KtorZ replied:
Type inference can most likely be resolved with a type annotation or application (getArg @TypeIExpect) .. Though, yeah, we don't need to spread the class more than necessary.

For the `ReaderT`, I don't think it provides any value here. An argument is really fine. Generally speaking, passing arguments around should be the preferred way over using a `Reader` or `ReaderT`. A reader is useful when multiple functions are dealing with some shared context and can help makes the code clearer. Here, for only 3 successive simple calls, I think it's overkill and adds unnecessary overhead.

KtorZ commented on app/launcher/Main.hs:
```diff
@@ -0,0 +1,102 @@
+    processes <- forM
+        [ nodeHttpBridgeOn nodePort network
+        , walletOn walletPort nodePort
+        ] createProcess
```
Side-note on style / readability, `mapM` is usually a better choice when your function is short and your iteree long. Basically, `forM = flip mapM`, so they are the same thing, but IMO, they read better depending on their arguments.

```hs
-- Iteree is short
forM xs $ \x -> do ...  -- Feels right
mapM (\x -> do ....) xs  -- Feels akward

-- Function is short
forM [many, different, things, because, there, is, no, intermediate, var] fn -- Feels akward
mapM fn [many, different, things, because, there, is no, intermediate, var] -- Feels right
```

This is really neat-picking though, I don't mind much as it is :+1:

Anviking replied:
👍 Thanks! Agreed.

Felt a bit wrong to have `createProcess` last, but never bothered to look it up.

KtorZ commented on app/launcher/Main.hs:
```diff
@@ -0,0 +1,102 @@
+walletOn :: Port "Wallet" -> Port "Node" -> CreateProcess
+walletOn w n = proc
+    "cardano-wallet-server"
+    ["--wallet-server-port", encode w, "--node-port", encode n]
```
:+1: Sweet.

KtorZ commented on app/launcher/Main.hs:
```diff
@@ -0,0 +1,102 @@
+            interruptProcessGroupOf h
+            cleanupProcess p
+
+nodeHttpBridgeOn :: Port "Node" -> Network -> CreateProcess
```
`Port (* :: Symbol)` --> Nice one :)

KtorZ commented on app/launcher/wallet.sh:
```diff
@@ -0,0 +1,6 @@
+
+
+#while true; do echo "I'm a wallet"; sleep 1; done
+echo "I'm a wallet"
```
If only it was that simple man...

Anviking committed:
Add Launcher

Anviking committed:
Add launcher tests

rvl committed:
Example of using withCreateProcess and race

It will ensure that:
- the subprocess is killed if the main thread dies.
- if the subprocess exits, the main thread will be cancelled.

KtorZ commented on cardano-wallet.cabal:
```diff
@@ -83,9 +83,8 @@ executable cardano-wallet-server
       NoImplicitPrelude
       OverloadedStrings
   ghc-options:
-      -threaded -rtsopts
```
We gotta need this :sweat_smile:  ... If the wallet server isn't threaded, it isn't going to do much serving requests.

Anviking replied:
Oops 😅 was supposed to be temporary

KtorZ committed:
Review launcher code, folder organization and API options

- allow launcher's command to take a 'setup' action before they're ran
  This allows for the wallet backend to wait a bit before it starts. Or,
  any kind of startup action we may think of, like generating TLS certs,
  etc ...

- review folder organization and tests scripts

- add Buildable intance to 'Command' and use that to format the initial info

- use --http-bridge-port instead of --node-port
